Fix: wrong baseDir (fixes #6450)

mysticatea opened the pull request:
Fixes #6450 

I'm not sure this fix is correct or not. I want to get any insights about `getBaseDir`.

And actually, this new test does not make sense. The test succeeds as well before this fix. The issue's error seems to be appeared only if eslint is installed to another package. I'm not sure how to write tests for this fix...

nzakas commented on lib/config/config-file.js:
```diff
@@ -315,6 +315,10 @@ function getBaseDir(configFilePath) {
     var projectPath = path.resolve(__dirname, "../../../");
 
+    if (path.basename(projectPath) === "node_modules") {
+        projectPath = path.dirname(projectPath);
```
Yikes! This seems like a fragile test, there has to be something else going on.

mysticatea commented:
I modified it, just add `../`.

**Before**

```
load( C:\Users\t-nagashima.AD\Documents\GitHub\sandbox\.eslintrc.json undefined undefined )
    resolvedPath: { filePath: 'C:\\Users\\t-nagashima.AD\\Documents\\GitHub\\sandbox\\.eslintrc.json' }
    dirname: C:\Users\t-nagashima.AD\Documents\GitHub\sandbox
    basedir: C:\Users\t-nagashima.AD\Documents\GitHub\sandbox\node_modules
    lookupPath: C:\Users\t-nagashima.AD\Documents\GitHub\sandbox\node_modules\node_modules
```

**After**

```
load( C:\Users\t-nagashima.AD\Documents\GitHub\sandbox\.eslintrc.json undefined undefined )
    resolvedPath: { filePath: 'C:\\Users\\t-nagashima.AD\\Documents\\GitHub\\sandbox\\.eslintrc.json' }
    dirname: C:\Users\t-nagashima.AD\Documents\GitHub\sandbox
    basedir: C:\Users\t-nagashima.AD\Documents\GitHub\sandbox
    lookupPath: C:\Users\t-nagashima.AD\Documents\GitHub\sandbox\node_modules
```

SimenB commented:
This fixes the problem for us (using `extends: './index.js'`)

ilyavolodin commented:
That looks much better then the check for `node_modules`.

nzakas commented:
I think we are heading in the wrong direction here. Adding `../` also seems very hacky to me because this worked fine before https://github.com/eslint/eslint/pull/6359 was merged. I think the correct path forward is to revert https://github.com/eslint/eslint/pull/6359 and try to solve https://github.com/eslint/eslint/issues/6358 a different way.

nzakas commented on lib/config/config-file.js:
```diff
@@ -313,7 +313,7 @@ function write(config, filePath) {
 
     // calculates the path of the project including ESLint as dependency
-    var projectPath = path.resolve(__dirname, "../../../");
+    var projectPath = path.resolve(__dirname, "../../../../");
```
This doesn't make sense to me. If this file hasn't moved, then why is this change necessary?

mysticatea replied:
Because it's wrong.
Do you think this `basedir` and `lookupPath` are correct?

```
load( C:\Users\t-nagashima.AD\Documents\GitHub\sandbox\.eslintrc.json undefined undefined )
    resolvedPath: { filePath: 'C:\\Users\\t-nagashima.AD\\Documents\\GitHub\\sandbox\\.eslintrc.json' }
    dirname: C:\Users\t-nagashima.AD\Documents\GitHub\sandbox
    basedir: C:\Users\t-nagashima.AD\Documents\GitHub\sandbox\node_modules
    lookupPath: C:\Users\t-nagashima.AD\Documents\GitHub\sandbox\node_modules\node_modules
```

ilyavolodin replied:
@mysticatea You might be right, in this case `basedir` doesn't look right. But with your fix, globally installed ESLint is failing even worse. It's now two directories off from where it should be looking (before this fix, it was one dir off).

mysticatea replied:
@ilyavolodin I confirmed the problem for globally installed ESLint.
So.... to fix this route cause, do we needs `=== "node_modules"` check as same as the first way? But it's rejected :confused:

mysticatea replied:
Mmm...
`basedir` is not the location of .eslintrc file for globally installed ESLint...

mysticatea replied:
I got it.
We cannot use `relativeTo` to resolve relative paths in .eslintrc files.
So #6468 is the correct way to fix this problem.

mysticatea replied:
I'm wondering that [this parser path](https://github.com/eslint/eslint/blob/master/lib/config/config-file.js#L515-L521) is working fine with this wrong `basedir` and `lookupPath`?

mysticatea replied:
I confirmed the parser path is not working.
We cannot use the `basedir` in order to resolve relative paths.

nzakas commented:
It also looks like the test isn't reproducing the bug that's being reported.

nzakas commented:
I think I figured it out here: https://github.com/eslint/eslint/pull/6468

mysticatea commented:
> It also looks like the test isn't reproducing the bug that's being reported.

Please see my first comment.

> And actually, this new test does not make sense. The test succeeds as well before this fix. The issue's error seems to be appeared only if eslint is installed to another package. I'm not sure how to write tests for this fix...

mysticatea commented:
I reverted the change of `projectRoot`.
The value of `getBaseDir()` is used for only resolving module path, the resolving searches in ancestor directories, so it's no problem.

Then, I found that [this parser path](https://github.com/eslint/eslint/blob/master/lib/config/config-file.js#L517) is not working as same as `extends` of `v2.13.0`.
I think the root cause is that we are using the value of `getBaseDir()` in order to resolve relative paths.
It's not the directory of the current file.

So I replaced `basedir` of both `parser` and `extends` by `dirname`.
`dirname` is the directory of the current file.

mysticatea commented:
I added some tests.

`basedir` is changed by whether `.eslintrc` exists in `projectPath` or not.
`tests/fixtures` directory exists in `projectPath`.
I added some tests which are executed outside of `projectPath`.

nzakas commented:
Ah! This looks much better. LGTM
